api/providers/tasks.py

```diff
@@ -563,10 +582,96 @@ def handle_registration_row(row, initiator, provider, schema, auto_approval=Fals
             registration.sanction.accept()
         except Exception as e:
             raise RegistrationBulkCreationRowError(
-                row.upload.id, row.id, row_title, row_external_id,
+                row.upload.id, row.id, title, external_id,
                 error=repr(e), approval_failure=True,
             )
         logger.info(f'Registration approved but pending moderation: [{registration._id}]')
+    return registration
+
+
+def bulk_upload_finish_job(upload, row_count, success_count, draft_errors, approval_errors, dry_run=True):
+    """Finish the bulk upload job: handle errors (if any), send outcome emails, and update job state.
+    """
+    provider = upload.provider
+    auto_approval = upload.provider.bulk_upload_auto_approval
+    initiator = upload.initiator
+
+    # All registration rows have failed draft creation
+    if len(draft_errors) == row_count:
+        upload.state = JobState.DONE_ERROR
+        message = f'All registration rows failed during bulk creation. ' \
+                  f'Upload ID: [{upload.id}], Draft Errors: [{draft_errors}]'
+        sentry.log_message(message)
+        logger.error(message)
+    # Some registration rows have failed draft creation
+    elif draft_errors:
+        upload.state = JobState.DONE_PARTIAL
+        if not approval_errors:
+            message = f'Some registration rows failed during bulk creation. ' \
+                      f'Upload ID: [{upload.id}]; Draft Errors: [{draft_errors}].'
+        else:
+            message = f'Some registration rows failed during bulk creation and some failed to be approved. ' \
+                      f'Upload ID: [{upload.id}]; Draft Errors: [{draft_errors}]; Approval Errors: [{approval_errors}]'
+        sentry.log_message(message)
+        logger.warning(message)
+    # All registration rows have finish the draft creation
+    else:
+        upload.state = JobState.DONE_FULL
+        if not approval_errors:
+            logger.info(f'All registration rows succeeded for bulk creation. Upload ID: [{upload.id}].')
+        else:
+            message = f'All registration rows succeeded for bulk creation but some have failed to be approved.' \
+                      f' Upload ID: [{upload.id}]; Approval Errors: [{approval_errors}]'
+            sentry.log_message(message)
+            logger.warning(message)
+            pass
+    # Reverse the error lists so that users see failed rows in the same order as the original CSV
+    draft_errors.reverse()
+    approval_errors.reverse()
```

Comment:
There is some issue with the order after parser update, will fix in another bug-fix PR.